cupy/cuda/compiler.py

```diff
@@ -605,77 +606,112 @@ def _convert_to_hip_source(source):
     ]
     for i, j in table:
         source = source.replace(i, j)
+    if not is_hiprtc:
+        return '#include <hip/hip_runtime.h>\n' + source
+
+    source = source.split('\n')
+    if extra_source is not None:
+        source = extra_source.split('\n') + source
+    source = [line for line in source if (
+        not line.startswith('#include')
+        and not line.startswith('#pragma once'))]
+    source = ('#include <hip/hip_runtime.h>\n'
+              + '#include <hip/hip_fp16.h>\n'
+              + '\n'.join(source))
 
-    return "#include <hip/hip_runtime.h>\n" + source
+    return source
 
 
 def _compile_with_cache_hip(source, options, arch, cache_dir, extra_source,
-                            backend='hiprtc', use_converter=True):
+                            backend='hiprtc', name_expressions=None,
+                            log_stream=None, cache_in_memory=False,
+                            use_converter=True):
     global _empty_file_preprocess_cache
+
+    if _is_cudadevrt_needed(options):
+        raise ValueError('separate compilation is not supported in HIP')
+
     if cache_dir is None:
         cache_dir = get_cache_dir()
     if arch is None:
         arch = os.environ.get('HCC_AMDGPU_TARGET')
         if arch is None:
             raise RuntimeError('HCC_AMDGPU_TARGET is not set')
     if use_converter:
-        source = _convert_to_hip_source(source)
+        source = _convert_to_hip_source(source, extra_source,
+                                        is_hiprtc=(backend == 'hiprtc'))
 
     env = (arch, options, _get_hipcc_version())
     base = _empty_file_preprocess_cache.get(env, None)
     if base is None:
         # This is checking of HIPCC compiler internal version
         base = _preprocess_hipcc('', options)
         _empty_file_preprocess_cache[env] = base
-    key_src = '%s %s %s %s' % (env, base, source, extra_source)
 
+    key_src = '%s %s %s %s' % (env, base, source, extra_source)
     key_src = key_src.encode('utf-8')
     name = '%s.hsaco' % hashlib.md5(key_src).hexdigest()
 
-    if not os.path.isdir(cache_dir):
-        try:
-            os.makedirs(cache_dir)
-        except OSError:
-            if not os.path.isdir(cache_dir):
-                raise
-
     mod = function.Module()
-    # To handle conflicts in concurrent situation, we adopt lock-free method
-    # to avoid performance degradation.
-    path = os.path.join(cache_dir, name)
-    if os.path.exists(path):
-        with open(path, 'rb') as file:
-            data = file.read()
-        if len(data) >= 32:
-            hash_value = data[:32]
-            binary = data[32:]
-            binary_hash = hashlib.md5(binary).hexdigest().encode('ascii')
-            if hash_value == binary_hash:
-                mod.load(binary)
-                return mod
+
+    if not cache_in_memory:
+        # Read from disk cache
+        if not os.path.isdir(cache_dir):
+            try:
+                os.makedirs(cache_dir)
```

Comment:
` exist_ok=True` to makedirs can be used to simplify this a bit?
(this is unrelated to the PR, but maybe we can clean it a bit

Reply:
Oh, sounds good. May I do this in the next PR? I plan to add a follow-up PR to cache the header source for hiprtc to avoid doing the string substation every time. It counts as enhancement instead of bug fix 😛 But if you like I can also do it here.

Reply:
NVM. I'm doing it now, stay tuned.

Reply:
Done in 1cdb828.